plugins/calm-vscode/src/extension.ts

```diff
@@ -0,0 +1,211 @@
+                vscode.commands.executeCommand('editor.action.formatDocument');
+                const editor = vscode.window.activeTextEditor;
+                try {
+                    const data: string = editor?.document.getText();
+                    const absPath: string = editor?.document?.uri?.fsPath;
+                    if (data && vscode.window.activeTextEditor) {
+                        diagnosticCollection.clear();
+                        const calmResult: ValidationOutcome = await callCalmValidate(data, absPath);
+                        const diagnostics: Diagnostic[] = calmResult.allValidationOutputs().map((calmValidation: ValidationOutput) => mapTo(data, calmValidation));
+                        diagnosticCollection.set(vscode.window.activeTextEditor.document.uri, diagnostics);
+                    }
+                } catch (error: unknown) {
+                    vscode.window.showErrorMessage('Error validating current file: ' + error);
+                }
+            }
+        });
+    });
+
+    context.subscriptions.push(validateCommand, visualizeCommand);
+    vscode.commands.executeCommand('calm-vscode.validate');
+}
+
+
+
+async function visualizeDocument(editor: vscode.TextEditor, panel: vscode.WebviewPanel, columnToShowIn: vscode.ViewColumn) {
+    try {
+        const data = editor?.document.getText();
+        const absPath = editor?.document?.uri?.fsPath;
+        if (data) {
+            const issueCount = await validationIssueCount(data, absPath);
+            if (issueCount > 0) {
+                vscode.window.showInformationMessage(`Please resolve the ${issueCount} validation issues before visualizing.`);
+                panel.dispose();
+                panel = undefined;
+            } else {
+                const svg = await visualize(data);
+                panel.webview.html = getWebviewContent(svg);
+                panel.reveal(columnToShowIn);
+            }
+        }
+    } catch (error: unknown) {
+        vscode.window.showErrorMessage('Error visualizing current file: ' + error);
+    }
+}
+
+function mapTo(jsonDataFromFile: string, calmValidation: ValidationOutput): vscode.Diagnostic {
+
+    let { line_start, line_end, character_start, character_end } = calmValidation;
+    if (calmValidation.line_start === undefined || (calmValidation.line_start === 1)) {
+        const data = getLineNumbers(jsonDataFromFile, calmValidation.path);
+        if (data.length === 1) {
+            line_start = data[0].line_start;
+            line_end = data[0].line_end;
+            character_start = data[0].character_start;
+            character_end = data[0].character_end;
+        }
+    }
+
+    return new Diagnostic(new Range(new vscode.Position(line_start, character_start), new vscode.Position(line_end, character_end)), calmValidation.message, convertSeverity(calmValidation.severity));
+}
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
```

Comment:
how come we need this? all variables look like they're being used to me

Reply:
Another fair spot, this was back when paths were hard-coded and `pageData` wasn't actually being used - it is now obviously!